File: FlutterProjects/Practice/hr_and_pms - v2 Upgrading/lib/features/leave/service/LeaveService.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:intl/intl.dart';

class LeaveService {
  final String baseUrl = 'http://localhost:8080/api/leaves';

  // Method to save a leave request
  Future<Map<String, dynamic>> saveLeaveRequest(Map<String, dynamic> leaveData) async {
    final response = await http.post(
      Uri.parse('$baseUrl/save'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode(leaveData),
    );

    if (response.statusCode == 200) {
      return jsonDecode(response.body);
    } else {
      throw Exception('Failed to save leave request');
    }
  }

  // Method to update a leave request
  Future<Map<String, dynamic>> updateLeaveRequest(int leaveId, Map<String, dynamic> leaveData) async {
    final response = await http.put(
      Uri.parse('$baseUrl/update/$leaveId'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode(leaveData),
    );

    if (response.statusCode == 200) {
      return jsonDecode(response.body);
    } else {
      throw Exception('Failed to update leave request');
    }
  }

  // Method to delete a leave request by ID
  Future<void> deleteLeave(int leaveId) async {
    final response = await http.delete(
      Uri.parse('$baseUrl/delete/$leaveId'),
    );

    if (response.statusCode != 204) {
      throw Exception('Failed to delete leave');
    }
  }

  // Method to get leave by ID
  Future<Map<String, dynamic>> getLeaveById(int leaveId) async {
    final response = await http.get(
      Uri.parse('$baseUrl/find/$leaveId'),
    );

    if (response.statusCode == 200) {
      return jsonDecode(response.body);
    } else {
      throw Exception('Leave not found');
    }
  }

  // Method to approve a leave request
  Future<Map<String, dynamic>> approveLeaveRequest(int leaveId) async {
    final response = await http.post(
      Uri.parse('$baseUrl/approve/$leaveId'),
    );

    if (response.statusCode == 200) {
      return jsonDecode(response.body);
    } else {
      throw Exception('Failed to approve leave request');
    }
  }

  // Method to reject a leave request
  Future<Map<String, dynamic>> rejectLeaveRequest(int leaveId) async {
    final response = await http.post(
      Uri.parse('$baseUrl/reject/$leaveId'),
    );

    if (response.statusCode == 200) {
      return jsonDecode(response.body);
    } else {
      throw Exception('Failed to reject leave request');
    }
  }

  // Method to get all pending leave requests
  Future<List<Map<String, dynamic>>> getPendingLeaveRequests() async {
    final response = await http.get(
      Uri.parse('$baseUrl/pending'),
    );

    if (response.statusCode == 200) {
      List<dynamic> data = jsonDecode(response.body);
      return List<Map<String, dynamic>>.from(data);
    } else {
      throw Exception('Failed to load pending leaves');
    }
  }

  // Method to get all leaves by leave type
  Future<List<Map<String, dynamic>>> getLeavesByType(String leaveType) async {
    final response = await http.get(
      Uri.parse('$baseUrl/type/$leaveType'),
    );

    if (response.statusCode == 200) {
      List<dynamic> data = jsonDecode(response.body);
      return List<Map<String, dynamic>>.from(data);
    } else {
      throw Exception('Failed to load leaves by type');
    }
  }

  // Method to get rejected leave requests
  Future<List<Map<String, dynamic>>> getRejectedLeaveRequests() async {
    final response = await http.get(
      Uri.parse('$baseUrl/rejected'),
    );

    if (response.statusCode == 200) {
      List<dynamic> data = jsonDecode(response.body);
      return List<Map<String, dynamic>>.from(data);
    } else {
      throw Exception('Failed to load rejected leaves');
    }
  }

  // Method to get approved leaves by user ID
  Future<List<Map<String, dynamic>>> getApprovedLeavesByUser(int userId) async {
    final response = await http.get(
      Uri.parse('$baseUrl/user/$userId/approved'),
    );

    if (response.statusCode == 200) {
      List<dynamic> data = jsonDecode(response.body);
      return List<Map<String, dynamic>>.from(data);
    } else {
      throw Exception('Failed to load approved leaves for user');
    }
  }

  // Method to get leaves by user and date range
  Future<List<Map<String, dynamic>>> getLeavesByUserAndDateRange(int userId, DateTime startDate, DateTime endDate) async {
    final response = await http.get(
      Uri.parse('$baseUrl/user/$userId/range?startDate=${DateFormat('yyyy-MM-dd').format(startDate)}&endDate=${DateFormat('yyyy-MM-dd').format(endDate)}'),
    );

    if (response.statusCode == 200) {
      List<dynamic> data = jsonDecode(response.body);
      return List<Map<String, dynamic>>.from(data);
    } else {
      throw Exception('Failed to load leaves by date range');
    }
  }

  // Method to get leaves by reason
  Future<List<Map<String, dynamic>>> getLeavesByReason(String reason) async {
    final response = await http.get(
      Uri.parse('$baseUrl/reason/$reason'),
    );

    if (response.statusCode == 200) {
      List<dynamic> data = jsonDecode(response.body);
      return List<Map<String, dynamic>>.from(data);
    } else {
      throw Exception('Failed to load leaves by reason');
    }
  }
}

File: FlutterProjects/Practice/API Backup/Backend with JWT- v5 AdvanceSalary/src/main/java/com/mrahmed/HRandPayrollManagementSystem/service/LeaveService.java
package com.mrahmed.HRandPayrollManagementSystem.service;

import com.mrahmed.HRandPayrollManagementSystem.entity.Leave;
import com.mrahmed.HRandPayrollManagementSystem.entity.LeaveType;
import com.mrahmed.HRandPayrollManagementSystem.entity.RequestStatus;
import com.mrahmed.HRandPayrollManagementSystem.repository.LeaveRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.util.List;
import java.util.Optional;

@Service
public class LeaveService {

    @Autowired
    private LeaveRepository leaveRepository;

    // Save leave request
    public Leave saveLeaveRequest(Leave leave) {
        int remainingDays = calculateRemainingLeaveDays(leave.getStartDate(), leave.getEndDate());
        leave.setRemainingLeave(remainingDays);
        return leaveRepository.save(leave);
    }

    //  update leave request
    public Leave updateLeaveRequest(Long id, Leave updatedLeave) {
        Optional<Leave> existingLeaveOpt = leaveRepository.findById(id);
        if (existingLeaveOpt.isPresent()) {
            Leave existingLeave = existingLeaveOpt.get();
            existingLeave.setStartDate(updatedLeave.getStartDate());
            existingLeave.setEndDate(updatedLeave.getEndDate());
            existingLeave.setRequestDate(updatedLeave.getRequestDate());
            existingLeave.setReason(updatedLeave.getReason());
            existingLeave.setRemainingLeave(updatedLeave.getRemainingLeave());
            existingLeave.setLeaveType(updatedLeave.getLeaveType());
            existingLeave.setRequestStatus(updatedLeave.getRequestStatus());
            existingLeave.setUser(updatedLeave.getUser());
            return leaveRepository.save(existingLeave);
        }
        throw new RuntimeException("Leave request not found with ID: " + id);
    }


    // Delete leave by ID
    public void deleteLeave(Long leaveId) {
        leaveRepository.deleteById(leaveId);
    }

    // Find leave by ID
    public Optional<Leave> getLeaveById(Long leaveId) {
        return leaveRepository.findById(leaveId);
    }

    // Approve a pending leave request
    public Leave approveLeaveRequest(Long leaveId) {
        Optional<Leave> leaveOpt = leaveRepository.findById(leaveId);
        if (leaveOpt.isPresent()) {
            Leave leave = leaveOpt.get();
            leave.setRequestStatus(RequestStatus.APPROVED);
            return leaveRepository.save(leave);
        }
        throw new RuntimeException("Leave request not found with ID: " + leaveId);
    }

    // Reject a leave request
    public Leave rejectLeaveRequest(Long leaveId) {
        Optional<Leave> leaveOpt = leaveRepository.findById(leaveId);
        if (leaveOpt.isPresent()) {
            Leave leave = leaveOpt.get();
            leave.setRequestStatus(RequestStatus.REJECTED);
            return leaveRepository.save(leave);
        }
        throw new RuntimeException("Leave request not found with ID: " + leaveId);
    }


    // Get all pending leave requests
    public List<Leave> getPendingLeaveRequests() {
        return leaveRepository.findPendingLeaveRequests();
    }


    // Get leaves by leave type
    public List<Leave> getLeavesByType(LeaveType leaveType) {
        return leaveRepository.findLeavesByType(leaveType);
    }

    //findRejectedLeaveRequests
    public List<Leave> getRejectedLeaveRequests() {
        return leaveRepository.findRejectedLeaveRequests();
    }

    // Get approved leaves for a specific user
    public List<Leave> getApprovedLeavesByUser(Long userId) {
        return leaveRepository.findApprovedLeavesByUser(userId);
    }

    // Get leaves for a specific user in a date range
    public List<Leave> getLeavesByUserAndDateRange(Long userId, LocalDate startDate, LocalDate endDate) {
        return leaveRepository.findLeavesByUserAndDateRange(userId, startDate, endDate);
    }

    // findLeavesByReason
    public List<Leave> getLeavesByReason(String reason) {
        return leaveRepository.findLeavesByReason(reason);
    }

    // Calculate remaining leaves based on start and end dates (helper method)
    public int calculateRemainingLeaveDays(LocalDate startDate, LocalDate endDate) {
        return (int) (endDate.toEpochDay() - startDate.toEpochDay());
    }

}
